manager/orchestrator/update/updater_test.go

```diff
@@ -592,3 +592,197 @@ func TestUpdaterOrder(t *testing.T) {
 		}
 	}
 }
+
+//  Tests special cases of Run() where slots contain multiple tasks.
+//  These cases are handled by the useExistingTask() function.
+func TestUpdaterUseExistingTask(t *testing.T) {
+	ctx := context.Background()
+	s := store.NewMemoryStore(nil)
+	assert.NotNil(t, s)
+	defer s.Close()
+
+	// Move tasks to their desired state.
+	watch, cancel := state.Watch(s.WatchQueue(), api.EventUpdateTask{})
+	defer cancel()
+	go func() {
+		for e := range watch {
+			task := e.(api.EventUpdateTask).Task
+			if task.Status.State == task.DesiredState {
+				continue
+			}
+			err := s.Update(func(tx store.Tx) error {
+				task = store.GetTask(tx, task.ID)
+				task.Status.State = task.DesiredState
+				return store.UpdateTask(tx, task)
+			})
+			assert.NoError(t, err)
+		}
+	}()
+
+	instances := 2
+	cluster := &api.Cluster{
+		// test cluster configuration propagation to task creation.
+		Spec: api.ClusterSpec{
+			Annotations: api.Annotations{
+				Name: store.DefaultClusterName,
+			},
+		},
+	}
+
+	service := &api.Service{
+		ID: "id1",
+		Spec: api.ServiceSpec{
+			Annotations: api.Annotations{
+				Name: "name1",
+			},
+			Mode: &api.ServiceSpec_Replicated{
+				Replicated: &api.ReplicatedService{
+					Replicas: uint64(instances),
+				},
+			},
+			Task: api.TaskSpec{
+				Runtime: &api.TaskSpec_Container{
+					Container: &api.ContainerSpec{
+						Image: "v:1",
+					},
+				},
+			},
+			Update: &api.UpdateConfig{
+				// avoid having Run block for a long time to watch for failures
+				Monitor: gogotypes.DurationProto(50 * time.Millisecond),
+			},
+		},
+	}
+
+	// 1: First test calls Run with one slot of two running tasks and
```

Comment:
nit: First test calls Run with one slot of two running tasks => First test creates two tasks for the same slot and calls updater.Run() and expects one task to be running and one to be shutdown.